virttest/qemu_devices/qbuses.py

```diff
@@ -650,6 +650,118 @@ def _increment_addr(self, addr, last_addr=None):
         return super(QPCIBus, self)._increment_addr(addr, last_addr=last_addr)
 
 
+class QPCIEBus(QPCIBus):
+    """
+    PCIE root bus representation (creates pcie-root-port for virtio pci
+    device by default.)
+    """
+
+    def __init__(self, busid, bus_type, aobject=None,
+                 root_type="pcie-root-port"):
+        """
+        :param busid: id of the bus (pcie.0)
+        :param bus_type: type of the bus (PCIE)
+        :param aobject: Related autotest object (pci.0)
+        :param root_type: root port type
+        """
+        super(QPCIEBus, self).__init__(busid, bus_type, aobject)
+        self.__root_ports = {}
+        self.__root_type = root_type
+
+    def is_direct_plug(self, device):
+        """
+        Justify if the device should be plug to pcie bus directly,
+        the device is not directly plugged only if it is a virtio
+        pci device and the device param direct_plug is "no".
+
+        :param device: the QBaseDevice object
+        :return: bool value for directly plug or not.
+        """
+        pcie_devices = ["virtio-net-pci", "virtio-blk-pci",
+                        "virtio-scsi-pci", "virtio-balloon-pci",
+                        "virtio-serial-pci", "virtio-rng-pci",
+                        "e1000e", "virtio-gpu-device"]
+        if device.get_param("driver") in pcie_devices and (
+                    device.get_param("direct_plug_pci", "no") == "no"):
```

Comment:
Seems that a device should be either a PCI device or PCI-E device except virtio devices, so I think the param `direct_plug_pci` can be only introduced to virtio devices, and also, I'd suggest using the name `pcie_hierarchy`.

Reply:
yep, something like `pcie_direct_plug` or `pcie_without_root_port` or just `pcie_directly` would be easier to find...